low-level/sensors/impl/centos_7/service_monitor.py

```diff
@@ -0,0 +1,392 @@
+        super(ServiceMonitor, self).__init__(self.SENSOR_NAME,
+                                                self.PRIORITY)
+
+        self.services_to_monitor = \
+            Conf.get(SSPL_CONF, f"{self.SERVICEMONITOR}>{self.MONITORED_SERVICES}", [])
+
+        self.not_active_services = {}
+        self.failed_services = []
+
+        self.service_status = {}
+
+        self.thread_sleep = \
+            int(Conf.get(SSPL_CONF, f"{self.SERVICEMONITOR}>{self.THREAD_SLEEP}", 1))
+
+        self.polling_frequency = \
+            int(Conf.get(SSPL_CONF, f"{self.SERVICEMONITOR}>{self.POLLING_FREQUENCY}", 15))
+
+        self.max_wait_time = \
+            int(Conf.get(SSPL_CONF, f"{self.SERVICEMONITOR}>{self.MAX_WAIT_TIME}", 60))
+
+    def read_data(self):
+        """Return the dict of service status."""
+        return self.service_status
+
+    def initialize(self, conf_reader, msgQlist, product):
+        """initialize configuration reader and internal msg queues."""
+
+        # Initialize ScheduledMonitorThread and InternalMsgQ
+        super(ServiceMonitor, self).initialize(conf_reader)
+
+        # Initialize internal message queues for this module
+        super(ServiceMonitor, self).initialize_msgQ(msgQlist)
+
+        # Integrate into the main dbus loop to catch events
+        DBusGMainLoop(set_as_default=True)
+
+        self._bus = SystemBus()
+        systemd = self._bus.get_object("org.freedesktop.systemd1", "/org/freedesktop/systemd1")
+        self._manager = Interface(systemd, dbus_interface='org.freedesktop.systemd1.Manager')
+
+        self.remove_disabled_services()
+
+        return True
+
+    def remove_disabled_services(self):
+        temp = self.services_to_monitor.copy()
+        for service in temp:
+            try:
+                if 'disabled' in str(self._manager.GetUnitFileState(service)):
+                    self.services_to_monitor.remove(service)
+            except DBusException:
+                self.services_to_monitor.remove(service)
+
+    def run(self):
+        logger.info(f"Monitoring Services : {self.services_to_monitor}")
+
+        temp_ser_to_mon = self.services_to_monitor.copy()
+        for service in temp_ser_to_mon:
+            err = self.connect_to_sig(service)
+            if err:
+                self.raise_alert(service, "N/A", "N/A", "N/A", "N/A",
+                    "N/A", "N/A", 0)
+                logger.error(f"{service} is not active initially. \n Error {err}")
+
+        logger.debug(f"failed_services : {self.failed_services}")
+        logger.debug(f"services_to_monitor : {self.services_to_monitor}")
+
+        # Retrieve the main loop which will be called in the run method
+        self._loop = gobject.MainLoop()
+
+        # Initialize the gobject threads and get its context
+        gobject.threads_init()
+        context = self._loop.get_context()
+
+        delay = 0
+        while self.is_running():
+            context.iteration(False)
```

Comment:
why not use scheduler thread class, like other sensors?

Also need try exception block around main logic. Align with other sensor code, avoid duplicated code & try to collect & refer from one place!

Reply:
Actually scheduler.enter() was not working because of the main_loop that was registered and iterated over continuously to catch the events, so instead of running main_loop, we are doing iteration over the context every second and checking for the event. also at the polling frequency, we are checking for the inactive list

Reply:
Added try exception.. 
Don't notice any duplicated code